ampd-handlers/src/bin/xrpl/handlers/multisig.rs

```diff
@@ -0,0 +1,379 @@
+
+use crate::error::Error;
+
+pub type Result<T> = error_stack::Result<T, Error>;
+
+#[derive(Clone, Debug, Deserialize)]
+#[try_from("wasm-signing_started")]
+pub struct SigningStartedEvent {
+    session_id: u64,
+    #[serde(deserialize_with = "deserialize_public_keys")]
+    pub_keys: HashMap<TMAddress, PublicKey>,
+    msg: MsgToSign,
+    expires_at: u64,
+    chain: ChainName,
+}
+
+fn deserialize_public_keys<'de, D>(
+    deserializer: D,
+) -> std::result::Result<HashMap<TMAddress, PublicKey>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let keys_by_address: HashMap<TMAddress, multisig::key::PublicKey> =
+        HashMap::deserialize(deserializer)?;
+
+    keys_by_address
+        .into_iter()
+        .map(|(address, pk)| Ok((address, pk.try_into().map_err(D::Error::custom)?)))
+        .collect()
+}
+
+#[derive(Debug, TypedBuilder)]
+pub struct Handler {
+    pub verifier: TMAddress,
+    pub multisig: TMAddress,
+    pub chain: ChainName,
+}
+
+impl Handler {
+    pub fn new(runtime: &HandlerRuntime, chain_name: ChainName) -> Self {
+        Self::builder()
+            .verifier(runtime.verifier.clone().into())
+            .multisig(runtime.contracts.multisig.clone().into())
+            .chain(chain_name)
+            .build()
+    }
+
+    fn submit_signature_msg(
+        &self,
+        session_id: impl Into<Uint64>,
+        signature: impl Into<HexBinary>,
+    ) -> MsgExecuteContract {
+        MsgExecuteContract {
+            sender: self.verifier.as_ref().clone(),
+            contract: self.multisig.as_ref().clone(),
+            msg: serde_json::to_vec(&ExecuteMsg::SubmitSignature {
+                session_id: session_id.into(),
+                signature: signature.into(),
+            })
+            .expect("submit signature msg should serialize"),
+            funds: vec![],
+        }
+    }
+}
+
+#[async_trait]
+impl EventHandler for Handler {
+    type Err = Error;
+    type Event = SigningStartedEvent;
+
+    async fn handle<HC: EventHandlerClient + Send + 'static>(
+        &self,
+        event: SigningStartedEvent,
+        client: &mut HC,
+    ) -> Result<Vec<Any>> {
+        let SigningStartedEvent {
+            session_id,
+            pub_keys,
+            msg,
+            expires_at,
+            chain,
+        } = event;
+
+        if !chain.eq(&self.chain) {
+            info!(
+                chain = chain.to_string(),
+                handler_chain = self.chain.to_string(),
+                "chain mismatch, skipping event"
+            );
+            return Ok(vec![]);
+        }
+
+        let latest_block_height = client
+            .latest_block_height()
+            .await
+            .change_context(Error::EventHandling)?;
+
+        if latest_block_height >= expires_at {
+            info!(
+                session_id = session_id.to_string(),
+                "skipping expired signing session"
+            );
+            return Ok(vec![]);
```

Comment:
we dont have a reusable version of this logic somewhere?

Reply:
we don't. At some point during the weekly meeting we discussed if we should abstract the multisig logic like we did with `voting.rs`, but the conclusion was that is not urgent since only xrpl has custom multisig logic. I would like to still do it if we have time though